### PR TITLE
Add ProjectModel server

### PR DESCRIPTION
**Pull request by troydai**
Round 1:

1) Port DTH;
2) Remove compilation related mechanism;
3) Port test cases

### CODE REVIEW

File: src/Microsoft.DotNet.ProjectModel.Server/ApplicationContext.cs [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+using Microsoft.DotNet.ProjectModel.Server.Helpers;
+using Microsoft.DotNet.ProjectModel.Server.InternalModels;
+using Microsoft.DotNet.ProjectModel.Server.Models;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.ProjectModel.Server
+{
+    public class ApplicationContext
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/ApplicationContext.cs]: Rename this, it should be a variant of `ProjectContext` like `RemoteProjectContext`

File: src/Microsoft.DotNet.ProjectModel.Server/ApplicationContext.cs [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+            if (_appPath.WasAssigned || _configure.WasAssigned || _filesChanged.WasAssigned || _refreshDependencies.WasAssigned)
+            {
+                bool triggerDependencies = _refreshDependencies.WasAssigned;
+
+                _appPath.ClearAssigned();
+                _configure.ClearAssigned();
+                _filesChanged.ClearAssigned();
+                _refreshDependencies.ClearAssigned();
+
+                state = _projectStateResolver.Resolve(_appPath.Value, _configure.Value, triggerDependencies, _remote.ProjectInformation?.ProjectSearchPaths);
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/ApplicationContext.cs]: Nooo, can we not bring any of this stuff back and just simply use the workspace context?

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            _loggerFactory = loggerFactory;
+            _protocolManager = new ProtocolManager(maxVersion: 4, loggerFactory: _loggerFactory);
+            _workspaceContext = WorkspaceContext.Create();
+            _frameworkReferenceResolver = new FrameworkReferenceResolver(ProjectContextBuilder.GetDefaultReferenceAssembliesPath());
+        }
+
+        public static int Main(string[] args)
+        {
+            var app = new CommandLineApplication();
+            app.Name = "dotnet-designtimeserver";
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: dotnet-projectmodel-server

File: src/Microsoft.DotNet.ProjectModel.Server/ProjectStateResolver.cs [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+using System.Linq;
+using Microsoft.DotNet.ProjectModel.Compilation;
+using Microsoft.DotNet.ProjectModel.Graph;
+using Microsoft.DotNet.ProjectModel.Server.Helpers;
+using Microsoft.DotNet.ProjectModel.Server.InternalModels;
+using Microsoft.DotNet.ProjectModel.Server.Models;
+
+namespace Microsoft.DotNet.ProjectModel.Server
+{
+    internal class ProjectStateResolver
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/ProjectStateResolver.cs]: This class does not need to exist.

File: test/Microsoft.DotNet.ProjectModel.Server.Tests/Helpers/TestHelper.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+{
+    public class TestHelper
+    {
+        private readonly string _tempPath;
+
+        public TestHelper()
+        {
+            LoggerFactory = new LoggerFactory();
+            LoggerFactory.AddConsole(LogLevel.Warning);
+            //LoggerFactory.AddConsole(LogLevel.Trace);
```
Comment (davidfowl) [inline test/Microsoft.DotNet.ProjectModel.Server.Tests/Helpers/TestHelper.cs]: Remove dead code

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+using System.Net.Sockets;
+using Microsoft.Dnx.Runtime.Common.CommandLine;
+using Microsoft.DotNet.ProjectModel.Resolution;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.ProjectModel.Server
+{
+    public class Program
+    {
+        private readonly Dictionary<int, ApplicationContext> _keepers = new Dictionary<int, ApplicationContext>();
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: keepers is such a bad name. How about projects

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        private Socket _listenSocket;
+
+        public Program(int intPort, string hostName, ILoggerFactory loggerFactory)
+        {
+            _port = intPort;
+            _hostName = hostName;
+            _loggerFactory = loggerFactory;
+            _protocolManager = new ProtocolManager(maxVersion: 4, loggerFactory: _loggerFactory);
+            _workspaceContext = WorkspaceContext.Create();
+            _frameworkReferenceResolver = new FrameworkReferenceResolver(ProjectContextBuilder.GetDefaultReferenceAssembliesPath());
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: Maybe just add a singleton instance of this. FrameworkReferenceResolver.Default that does this

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            app.Description = ".NET Design Time Server";
+            app.HelpOption("-?|-h|--help");
+
+            app.OnExecute(() =>
+            {
+                app.ShowHelp();
+                return 0;
+            });
+
+            app.Command("run", cla =>
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: Why is there a run?

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs (hunk 3 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: Can we have a better name?

File: src/Microsoft.DotNet.ProjectModel.Server/ProjectStateResolver.cs (hunk 4 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/ProjectStateResolver.cs]: It's not that simple. Through `ProjectContext` represents most of the data to diff, it still needs a scan of the diagnostics and libraries to build up the actual data model for returning data. For example: additional diagnostics are created when a dependencies flip from project to package. 

Some of the data transform can be delayed to the point when actual response is determined to be required. I'll move them to independent logic.

**Comment by troydai** [comment]
Updated. Ping @davidfowl

File: Microsoft.DotNet.Cli.sln [hunk 9]
```diff
@@ -57,9 +57,15 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "LoadContextTest", "test\Loa
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.DotNet.Tools.New", "src\Microsoft.DotNet.Tools.New\Microsoft.DotNet.Tools.New.xproj", "{BC765FBF-AD7A-4A99-9902-5540C5A74181}"
 EndProject
+<<<<<<< HEAD
```
Comment (davidfowl) [inline Microsoft.DotNet.Cli.sln]: What?

File: src/Microsoft.DotNet.ProjectModel.Server/ConnectionContext.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            _hostName = hostName;
+            _projectContextManagers = projectContextManagers;
+
+            _queue = new ProcessingQueue(new NetworkStream(acceptedSocket), loggerFactory);
+            _queue.OnReceive += message =>
+            {
+                // Enumerates all project contexts and return them to the
+                // sender
+                if (message.MessageType == MessageTypes.EnumerateProjectContexts)
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/ConnectionContext.cs]: Remove this code it's no longer required

File: src/Microsoft.DotNet.ProjectModel.Server/Helpers/FlipDependencyFinder.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.ProjectModel.Graph;
+
+namespace Microsoft.DotNet.ProjectModel.Server.Helpers
+{
+    internal class FlipDependencyFinder
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Helpers/FlipDependencyFinder.cs]: Can this just be a helper method in one of the existing classes?

Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Helpers/FlipDependencyFinder.cs]: Don't bother, it just needs a better name

File: src/Microsoft.DotNet.ProjectModel.Server/InternalModels/ProjectInfo.cs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        public ProjectInfo(ProjectContext context,
+                           string configuration,
+                           IEnumerable<string> currentSearchPaths)
+        {
+            _allExports = context.CreateExporter(configuration).GetAllExports().ToList();
+            _allDiagnostics = context.LibraryManager.GetAllDiagnostics();
+
+            Context = context;
+            Configuration = configuration;
+
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/InternalModels/ProjectInfo.cs]: Can you unlinq this code and write a simple loop that computes everything?

File: src/Microsoft.DotNet.ProjectModel.Server/MessageTypes.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        public const string ProjectContexts = nameof(ProjectContexts);
+        public const string Initialize = nameof(Initialize);
+        public const string ChangeConfiguration = nameof(ChangeConfiguration);
+        public const string RefreshDependencies = nameof(RefreshDependencies);
+        public const string RestoreComplete = nameof(RestoreComplete);
+        public const string FilesChanged = nameof(FilesChanged);
+        public const string GetDiagnostics = nameof(GetDiagnostics);
+        public const string ProtocolVersion = nameof(ProtocolVersion);
+
+        // Outcomming
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/MessageTypes.cs]: Outgoing

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+            while (true)
+            {
+                var acceptSocket = _listenSocket.Accept();
+                logger.LogInformation($"Client accepted {acceptSocket.LocalEndPoint}");
+
+                var connection = new ConnectionContext(_hostName,
+                                                       acceptSocket,
+                                                       _protocolManager,
+                                                       _workspaceContext,
+                                                       _projectContextManagers,
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: What's up with these random spaces

File: src/Microsoft.DotNet.ProjectModel.Server/MessageTypes.cs (hunk 13 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/MessageTypes.cs]: ... yeah

File: src/Microsoft.DotNet.ProjectModel.Server/InternalModels/ProjectInfo.cs (hunk 12 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/InternalModels/ProjectInfo.cs]: ok... but this looks wonderful.

File: src/Microsoft.DotNet.ProjectModel.Server/Helpers/FlipDependencyFinder.cs (hunk 11 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/Helpers/FlipDependencyFinder.cs]: suggestion?

File: src/Microsoft.DotNet.ProjectModel.Server/ConnectionContext.cs (hunk 10 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/ConnectionContext.cs]: my pleasure.

File: Microsoft.DotNet.Cli.sln (hunk 9 above)
Comment (troydai) [inline Microsoft.DotNet.Cli.sln]: ugr.. will fix.

File: src/Microsoft.DotNet.ProjectModel.Server/Program.cs (hunk 14 above)
Comment (troydai) [inline src/Microsoft.DotNet.ProjectModel.Server/Program.cs]: there is no random space here. `acceptSocket` doesn't start with `_`, just an illusion.

**Comment by troydai** [comment]
:watch:

File: scripts/compile.ps1 [hunk 15]
```diff
@@ -50,13 +50,19 @@ Download it from https://www.cmake.org
...
     # Restore packages
     header "Restoring packages"
-    & "$DnxRoot\dnu" restore "$RepoRoot" --quiet --runtime "osx.10.10-x64" --runtime "ubuntu.14.04-x64" --runtime "win7-x64" --no-cache
+    & "$DnxRoot\dnu" restore "$RepoRoot\src" --quiet --runtime "osx.10.10-x64" --runtime "ubuntu.14.04-x64" --runtime "win7-x64" --no-cache
+    & "$DnxRoot\dnu" restore "$RepoRoot\test" --quiet --runtime "osx.10.10-x64" --runtime "ubuntu.14.04-x64" --runtime "win7-x64" --no-cache
+
+    $oldErrorAction=$ErrorActionPreference
+    $ErrorActionPreference="SilentlyContinue"
+    & "$DnxRoot\dnu" restore "$RepoRoot\testapp" --quiet --runtime "osx.10.10-x64" --runtime "ubuntu.14.04-x64" --runtime "win7-x64" --no-cache
+    $ErrorActionPreference=$oldErrorAction
```
Comment (anurse) [inline scripts/compile.ps1]: Eeeew. Who told you to do this? Oh right... me. Carry on then.

Comment (troydai) [inline scripts/compile.ps1]: Desperate time, desperate measure.

File: scripts/compile.sh [hunk 16]
```diff
@@ -48,7 +48,11 @@ DOTNET_PATH=$(which dotnet)
 PREFIX="$(cd -P "$(dirname "$DOTNET_PATH")/.." && pwd)"
 
 header "Restoring packages"
-$DNX_ROOT/dnu restore "$REPOROOT" --quiet --runtime "$RID" --no-cache
+$DNX_ROOT/dnu restore "$REPOROOT/src" --quiet --runtime "$RID" --no-cache
```
Comment (brthor) [inline scripts/compile.sh]: Why break this up into 3 restores?

Comment (troydai) [inline scripts/compile.sh]: Need to escape the error during restore _testapp_. There are projects in _testapp_ intentionally have broken dependencies for test. The restore failure on it shouldn't fail the build.

Comment (brthor) [inline scripts/compile.sh]: Makes sense. Thanks!

**Comment by anurse** [comment]
@brthor I'm not sure what the build failure is. Does it look familiar to you? It seems vaguely related to the project.json validator you added.

**Comment by brthor** [comment]
@Sridhar-MS 
`m[31m/tmp/DotnetCLITests/7b439a20-0fa8-4d42-b92d-3e13802d538a/Program.cs(12,18): error CS1002: ; expected[39m[22m`

This looks like it's related to the Temp Directory infrastructure? Any ideas?

**Comment by Sridhar-MS** [comment]
All the publish tests are indeed passing
> 16:17:04 === TEST EXECUTION SUMMARY ===
16:17:04    Microsoft.DotNet.Tools.Publish.Tests  Total: 9, Errors: 0, Failed: 0, Skipped: 0, Time: 97.365s

There is one negative [test](https://github.com/dotnet/cli/blob/master/test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs#L135) in which compilation fails and the test also expect publish to fail. And it works as expected. But jenkins tries to parse the build output and report any common issues.

But the actual failure is in pjvalidator
> 'D:\j\workspace\dotnet_cli\debug_windows_nt_prtest\tools\MultiProjectValidator' cannot be published  for '<no framework provided>' 'win7-x64'

**Comment by brthor** [comment]
@troydai Are you doing a restore for the `tools` folder? I expect this may be the issue here.

**Comment by troydai** [comment]
@Sridhar-MS thanks for the explanation. any suggest how to fix it? should we configure the Jenkins to ignore the stderr or suppress the test? and any reason why it doesn’t fail on the master branch?

**Comment by troydai** [comment]
@brthor No, I didn’t. I change the restore to specific target three folder: `src`, `test` and `testapp`. And in order to suppress the error message, I redirect both stdout and stderr to `null` when restore `testapp` folder. See here: https://github.com/dotnet/cli/blob/trdai/dth/scripts/compile.ps1#L63.

**Comment by troydai** [comment]
@brthor actually, I think publish MultiProjectValidator without not restore tools folder is the actually problem. https://github.com/dotnet/cli/blob/trdai/dth/scripts/compile.ps1#L162

File: src/Microsoft.DotNet.ProjectModel.Server/ProjectContextManager.cs [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+
+                state = ProjectState.Create(_appPath.Value, _configure.Value, _workspaceContext, _remote.ProjectSearchPaths);
+            }
+
+            if (state == null)
+            {
+                return false;
+            }
+
+            var frameworkReferenceResolver = FrameworkReferenceResolver.Default;
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel.Server/ProjectContextManager.cs]: Why is this here?

File: src/Microsoft.DotNet.ProjectModel/Resolution/FrameworkReferenceResolver.cs [hunk 18]
```diff
@@ -33,6 +36,19 @@ public FrameworkReferenceResolver(string referenceAssembliesPath)
...
         
         public string ReferenceAssembliesPath { get; }
+        
+        public static FrameworkReferenceResolver Default
+        {
+            get
+            {
+                if (_default == null)
+                {
+                    _default = new FrameworkReferenceResolver(ProjectContextBuilder.GetDefaultReferenceAssembliesPath());
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/Resolution/FrameworkReferenceResolver.cs]: I would move GetDefaultReferenceAssembliesPath to this type as well.

**Comment by davidfowl** [comment]
I have a bunch of comments about the Messenger<T> pattern but lets go with this for now so we can iterate. :shipit:

**Comment by troydai** [comment]
I'll rebase and merge once it passes on CI.

**Comment by troydai** [comment]
@brthor @Sridhar-MS the build fails because of 
```
Unable to locate Dependency [1m[31mMicrosoft.CodeAnalysis.CSharp[39m[22m 1.2.0-beta-20151215-01
``
I'll replace it with build 1.2.0-beta-20151224-03.

**Comment by troydai** [comment]
I'll merge to main in minutes.
